fix: Server crashes due to insufficient schema validation when schema type is not a string

rgunindi opened the pull request:
Added a type check in mongoFieldToParseSchemaField to ensure `type` is a string before calling `startsWith`. This prevents crashes when Parse Server processes MongoDB schema fields with undefined, null, or unexpected type values.

Closes #9847

## Pull Request

- Report security issues [confidentially](https://github.com/parse-community/parse-server/security/policy).
- Any contribution is under this [license](https://github.com/parse-community/parse-server/blob/alpha/LICENSE).
- Link this pull request to an [issue](https://github.com/parse-community/parse-server/issues/9847).

## Issue
<!-- Add the link to the issue that this PR closes. -->
- Link to issue [issue](https://github.com/parse-community/parse-server/issues/9847)



## Approach
<!-- Describe the changes in this PR. -->

## Tasks
<!-- Delete tasks that don't apply. -->

- [x] Add tests


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Improved schema validation for Mongo-backed storage: invalid, empty, or non-string field types now produce clear, user-facing errors instead of crashes.

* **Tests**
  * Added comprehensive tests covering valid field-type mappings and verifying invalid inputs reliably throw the expected errors.

* **Chores**
  * Strengthened input guards to reduce ambiguous failures and improve reliability during schema setup and migrations.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

rgunindi committed:
fix: Prevent TypeError when type is not a string

Added a type check in mongoFieldToParseSchemaField to ensure `type` is a string
before calling `startsWith`. This prevents crashes when Parse Server processes
MongoDB schema fields with undefined, null, or unexpected type values.

Closes #9847

parseplatformorg commented:
### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/acinader/pr-checks/c2473b92-caa8-47da-b036-770e45a4e1a0?refs=5852d3c32ff669a000ee3d65933af7f9450268f9&source=prChecksComment)

mtrezza reviewed (commented):
Please add a test to demonstrate the issue. You experienced this in Parse Server 6, we also want to make sure this bug still exists in PS 8.

rgunindi committed:
test: Add comprehensive tests for mongoFieldToParseSchemaField type validation and clear error message

- Add tests for valid field type conversions
- Add tests for invalid type handling (null, undefined, non-string values)
- Ensure Parse.Error code is thrown for invalid inputs
- Test coverage for the type validation improvements
- Add a clear error explanation

rgunindi commented:
> Please add a test to demonstrate the issue. You experienced this in Parse Server 6, we also want to make sure this bug still exists in PS 8.

@mtrezza 
Thanks for the suggestion. I’ve added tests that demonstrate the issue and verified them against also current version ps8.

<img width="970" height="337" alt="image" src="https://github.com/user-attachments/assets/d07235c9-b806-4b31-ac0b-c54e2c4c9fd1" />

mtrezza commented:
Thanks, did you mean that you verified that the bug exists in PS8?

mtrezza commented on src/Adapters/Storage/Mongo/MongoSchemaCollection.js:
```diff
@@ -2,6 +2,14 @@ import MongoCollection from './MongoCollection';
+  if (!type || typeof type !== 'string') {
+    throw new Parse.Error(
+      Parse.Error.INVALID_SCHEMA_OPERATION,
+      `Invalid field type: ${type}. Expected a string. Field type must be one of: string, number, boolean, date, map, object, array, geopoint, file, bytes, polygon, or a valid relation/pointer format.`
```
I would not list all the types here; devs can consult docs. Easier to maintain if anything changes in the future, e.g. new type.

rgunindi replied:
ok, I just use `Invalid field type: ${type}.`

rgunindi replied:
`Invalid field type: ${type}. Expected a string. Fix the type mismatch in your schema configuration.`

How is this ?

mtrezza commented on spec/MongoSchemaCollectionAdapter.spec.js:
```diff
@@ -1,5 +1,6 @@
 'use strict';
 
+const Parse = require('parse/node');
```
Is this needed?

rgunindi replied:
Actually, I used it directly to verify that there was a parse error. I will change to error code.

rgunindi commented:
> Thanks, did you mean that you verified that the bug exists in PS8?

Yes, of course. This bug exists in Ps8.

Moumouls commented on src/Adapters/Storage/Mongo/MongoSchemaCollection.js:
```diff
@@ -2,6 +2,14 @@ import MongoCollection from './MongoCollection';
+  if (!type || typeof type !== 'string') {
+    throw new Parse.Error(
+      Parse.Error.INVALID_SCHEMA_OPERATION,
+      `Invalid field type: ${type}. Expected a string. Fix the type mismatch in your schema configuration.`
```
question: to help developer to identity the field, any chance to mention the impacted field and current class name ?

rgunindi replied:
Yes much more sense. Now when there's a type validation error, developers will see which field and class is causing the issue.

mtrezza committed:
Merge branch 'alpha' into fix/type-validation-mongo-schema

rgunindi committed:
fix: Improve error message in mongoFieldToParseSchemaField with field and class details

- Add fieldName and className parameters to mongoFieldToParseSchemaField function
- Update error message to include specific field name and class information
- Show collection name (_SCHEMA) and document ID for better debugging
- Pass field name and class ID from mongoSchemaFieldsToParseSchemaFields
- Enhanced error message format: 'Invalid field type: X for field Y in class _SCHEMA (id: Z)'
- Improves developer experience by providing clear context for schema validation errors

Moumouls commented on spec/MongoSchemaCollectionAdapter.spec.js:
```diff
@@ -160,6 +160,7 @@ describe('MongoSchemaCollection', () => {
       } catch (error) {
         expect(error.code).toBe(255);
         expect(error.message).toContain('Invalid field type');
+        expect(error.message).toContain('Expected a string');
```
Issue: When using contain, you can either expect the full error message with equal, or add a contain on the class name and impacted field name to correctly cover the error message 😃.

This way, your feature will not be broken in case of future contributions that touch this area.

rgunindi replied:
I think using toContain makes more sense here, since the error message often includes extra details (like className, fieldName, etc.). A strict would be more fragile.
toContain checks specifically for the className and fieldName would validate the important parts of the message without making the test overly strict.
If this doesn't make sense to you, I can change it.

Moumouls replied:
Hi @rgunindi, the className and field name are important in this kind of error, so I think it’s a good idea to cover them.

This part of the code won’t change often, so having a strict, fragile test is okay. If it becomes too tedious to maintain, it can always be switched back to a simple contain, but I believe it will work fine as is 🙂
